src/pipeline.h

```diff
@@ -457,14 +482,40 @@ class Pipeline {
         return static_cast<uint32_t>(i);
       }
     }
+    assert(0 && "GetShaderGroupIndex called too early");
+
     return static_cast<uint32_t>(-1);
   }
-
-  /// Retrieves a list of all Shader Groups.
-  const std::vector<std::unique_ptr<ShaderGroup>>& GetShaderGroups() const {
+  /// Remembers number of shader groups belonging to this pipeline without
+  /// libraries addition.
+  void SetNonLibShaderGroupCount() {
+    assert(non_lib_shader_groups_count_ == -1); // Should be set only once
+    non_lib_shader_groups_count_ = shader_groups_.size();
+  }
+  /// Returns number of shader groups belonging to this pipeline without
+  /// libraries addition.
+  size_t GetNonLibShaderGroupCount() {
```

Comment:
As I have written above:

> On practical stage: amber file in pipeline (that uses some pipeline library) can refer to shader group names inherited from pipeline libraries used, thus we need to **either parse whole tree of used pipeline libraries or do as I just did**: include pipeline groups into pipeline that uses pipeline library. Same with shaders (pStages).

But to create pipeline I need to know how many stages and groups are this pipeline native.

Reply:
It seems like it would be similer to keep the tree of pipelines and assemble what we need when we create the pipeline. Keep the create script closer to the actual amber script instead of attempting to store some compiled pipeline information.